### PR TITLE
✳️ Visualization Use Case Implementation using clean Architecture

### LINKED ISSUE #6: Visualization Implementation - Implement clean architecture, and week, month and year use cases.
Data visualization to track user's habit completion over a time period.

### PR DESCRIPTION
**Pull request by MadhavKanna**
Description: This PR implements the visualization use case, complete with all layers required for successful functionality. PR complies fully with clean architecture. PR Implemented Controller, InputBoundary, Interactor, DsGateway, Database component, OutputBoundary, and Presenter. 

- [ ] Unit tests (For task feature): Before PR: 17% After PR: 50%
- [x] Exception Handling: Exceptions are handled.
- [x] JavaDocs Documentation
- [x] SOLID principles
- [x] Clean Architecture: Aligned with CA.

### CODE REVIEW

File: src/main/java/use_cases/visualization/visualization_month/VisualizationMonthRequestModel.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Request model passed from controller to Inputboundary interface, implemented by interactor
+ */
+public class VisualizationMonthRequestModel {
+    private String habitName;
+    private Date startDate;
+
+    /**
+     * Create a request model for visualize month use case
+     * @param habitName
```
Comment (AngelEyesMagicMaker) [inline src/main/java/use_cases/visualization/visualization_month/VisualizationMonthRequestModel.java:15]: Maybe explain the parameter a little bit.

File: src/main/java/use_cases/visualization/visualization_month/VisualizationMonthResponseModel.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        this.habitName = habitName;
+        this.startDate = startDate;
+    }
+
+    // setter and getter methods
+    String getHabitName(){return this.habitName;}
+
+    Date getStartDate(){return this.startDate;}
+
+    void setHabitName(String habitName){this.habitName = habitName;}
```
Comment (AngelEyesMagicMaker) [inline src/main/java/use_cases/visualization/visualization_month/VisualizationMonthResponseModel.java:25]: May add Javadoc for the setter methods.

File: src/main/java/use_cases/visualization/visualization_week/VisualizationWeekDsGateway.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    /**
+     * check that the habit exists for the current user
+     * @param habitName
+     * @return
+     */
+    boolean habitExistsByName(String habitName);
+
+    /**
+     * check that there is enough data to meaningfully visualize data. Enough data is
+     * characterised by presence of at least one record of compleion of habit
```
Comment (AngelEyesMagicMaker) [inline src/main/java/use_cases/visualization/visualization_week/VisualizationWeekDsGateway.java:24]: Good explanation, add explanations for what will return for the methods.

**Review by EmilyDo — DISMISSED** [review]

**Review by AngelEyesMagicMaker — DISMISSED** [review]

**Review by AngelEyesMagicMaker — APPROVED** [review]

**Review by m-banthwan — APPROVED** [review]
approved!

File: src/main/java/use_cases/visualization/visualization_month/VisualizationMonthOutputBoundary.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+     * return response model for successful use case execution
+     * @param visualizationMonthResponseModel
+     * @return
+     */
+    VisualizationMonthResponseModel prepareSuccessView(VisualizationMonthResponseModel visualizationMonthResponseModel);
+
+    /**
+     * Return response model for unsuccessful use case execution
+     * @param errorMessage
+     * @return
```
Comment (JoAndNight) [inline src/main/java/use_cases/visualization/visualization_month/VisualizationMonthOutputBoundary.java:15]: You need to tell what is returned

File: src/main/java/controllers/visualization/VisualizationWeekController.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+import java.util.Calendar;
+import java.util.Date;
+
+public class VisualizationWeekController {
+
+    final VisualizationWeekInputBoundary visualizationInput;
+
+    public VisualizationWeekController(VisualizationWeekInputBoundary visualizationInput){
+        this.visualizationInput = visualizationInput;
```
Comment (m-banthwan) [inline src/main/java/controllers/visualization/VisualizationWeekController.java:16]: Please include javadoc after implementation in your next PR

File: src/main/java/database/DatabaseVisualization.java [hunk 6]
```diff
@@ -1,4 +1,117 @@
...
+    /**
+     *  Create chart for week-wise visualization
+     * @param requestModel - Ds RequestModel used to store data required to create week-wise visualization
+     */
+    @Override
+    public void createChart(VisualizationWeekDsRequestModel requestModel) {
+
+        ArrayList<Habit> habitList = new ArrayList();
+        HashMap<String, Integer> fullHabitData = this.collection.get(requestModel.getHabitName()).getFrequencyMap();
+        SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy");
```
Comment (EmilyDo) [inline src/main/java/database/DatabaseVisualization.java:40]: variable name not descriptive

File: src/main/java/use_cases/visualization/visualization_week/VisualizationWeekDsRequestModel.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package use_cases.visualization.visualization_week;
+
+import java.util.Calendar;
+
+public class VisualizationWeekDsRequestModel {
+    private String habitName;
+    private Calendar startDate;
+    private String imagePath;
+
+    public VisualizationWeekDsRequestModel(String habitName, Calendar startDate, String imagePath) {
```
Comment (m-banthwan) [inline src/main/java/use_cases/visualization/visualization_week/VisualizationWeekDsRequestModel.java:10]: Please Add javadoc to this whole file

File: src/main/java/use_cases/visualization/visualization_month/VisualizationMonthResponseModel.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+     */
+    public VisualizationMonthResponseModel(String habitName, Date startDate, String imagePath) {
+        this.habitName = habitName;
+        this.startDate = startDate;
+    }
+
+    // setter and getter methods
+    String getHabitName(){return this.habitName;}
+
+    Date getStartDate(){return this.startDate;}
```
Comment (JoAndNight) [inline src/main/java/use_cases/visualization/visualization_month/VisualizationMonthResponseModel.java:23]: You need to add more specific JavaDoc

File: src/main/java/use_cases/visualization/visualization_week/VisualizationWeekInteractor.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+public class VisualizationWeekInteractor implements VisualizationWeekInputBoundary {
+
+    final VisualizationWeekDsGateway visualizationWeekDsGateway;
+    final VisualizationWeekOutputBoundary visualizationWeekPresenter;
+
+    public VisualizationWeekInteractor(VisualizationWeekDsGateway visualizationWeekDsGateway, VisualizationWeekOutputBoundary visualizationWeekPresenter) {
+        this.visualizationWeekDsGateway = visualizationWeekDsGateway;
+        this.visualizationWeekPresenter = visualizationWeekPresenter;
+    }
+
```
Comment (m-banthwan) [inline src/main/java/use_cases/visualization/visualization_week/VisualizationWeekInteractor.java:12]: Please add javadoc to Interactor in your next PR

**Review by JoAndNight — APPROVED** [review]

**Review by m-banthwan — APPROVED** [review]

**Review by EmilyDo — APPROVED** [review]

**Pull request merged**